document-portal/document-portal.c

```diff
@@ -1416,6 +1416,68 @@ portal_list (GDBusMethodInvocation *invocation,
   return TRUE;
 }
 
+static gboolean
+portal_get_real_path (GDBusMethodInvocation *invocation,
+             GVariant *parameters,
+             XdpAppInfo *app_info)
+{
+  const char *id = NULL;
+  g_autoptr(PermissionDbEntry) entry = NULL;
+
+  g_variant_get (parameters, "(&s)", &id);
+
+  XDP_AUTOLOCK (db);
+
+  entry = permission_db_lookup (db, id);
+
+  if (!entry)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                                             "Invalid ID passed");
+      return TRUE;
+    }
+
+  if (!xdp_app_info_is_host (app_info))
+    {
+      g_autofree const char **apps = NULL;
+      g_autofree const char *app_id = NULL;
+      gboolean id_found = FALSE;
+      int i;
+
+      app_id = xdp_app_info_get_id(app_info);
+
+      apps = permission_db_entry_list_apps (entry);
+      for (i = 0; apps[i] != NULL; i++)
+        {
+          if (g_strcmp0 (app_id, apps[i]) == 0)
+            {
+              id_found = TRUE;
+              break;
+            }
+        }
+
+      if (!id_found)
+        {
+          g_dbus_method_invocation_return_error (invocation,
+                                                 XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                                                 "Invalid ID passed");
```

Comment:
Is it possible to have a ID part of a Flatpak, but with no permission?

Reply:
What I mean is, that you already found a database entry for given ID, right? At this point (in the code) you are looking whether the database entry has entry for your application (based on app_id). Maybe the variable name `id_found` is a bit confusing and it should be `app_found` perhaps? If the app is not found in the database entry, I think it means you don't have permission to this document and you shouldn't be able to get the real path. That's my understanding of this code, therefore the error shouldn't be about a wrong argument, because that's what you already checked at the beginning of this function, but  a missing entry for given `app_id` is a problem and means you shouldn't be able to get the real path for missing permission.

Reply:
I use the same error on both places, so a sandboxed App don't know if a ID does not exists or if the App don't have Permission for the ID.  I don't think it's a security issue if the App knows a ID exists, but I think Sandboxed Apps don't need to know about that.

Reply:
As a client you should get the real reason why your request failed. I don't want to debug why my request failed when providing an existing document ID, while the reason for failure is that I no longer have permission to this document. There are [other calls](https://github.com/flatpak/xdg-desktop-portal/blob/main/document-portal/document-portal.c#L176) in this API giving you information about you not having enough permission.